### DISCUSSION TITLE
Why are Required Actions shown in the middle of password reset flow?

### DISCUSSION THREAD

**Discussion by mrpatrickpilch** [keycloak/keycloak#16527] [Q&A] [answered]
If the user has a pending required action (an example for the purpose of this discussion - Terms and Conditions), why does it appear as the first step in the password reset flow? It's jarring - you click a hyperlink in an email and are presented with an entirely unrelated step/page.

I'm using required actions for additional "steps" in user registration, because creating a multi-step registration process is fraught with complication. After submitting a password and the `User Registration` form action completes successfully, the user has a valid profile and they can abandon the website before completing all the required actions. If they come back at a later point in time having forgotten their password, the password reset flow drops them into the next required action. This doesn't feel right.

On a related note, it appears as though resetting the password does not set the email to verified either. In `org.keycloak.authentication.authenticators.resetcred.ResetCredentialEmail#action`, there is `context.getUser().setEmailVerified(true);` which leads me to believe it should. But after completing the required actions, resetting the password, and logging in - the email is not verified. (At least on Keycloak 20.0.1)

**Comment by mrpatrickpilch** [comment]
Documenting investigation for any others who may be facing a similar issue:

Seems like the crux is due to `org.keycloak.services.managers.AuthenticationManager#nextRequiredAction` (being called by `org.keycloak.authentication.actiontoken.resetcred.ResetCredentialsActionTokenHandler.ResetCredsAuthenticationProcessor#authenticationComplete`) first prioritizing required actions that are set on the *UserModel*, and _then_ moving along to required actions set on the *AuthenticationSessionModel*.  Because `org.keycloak.authentication.authenticators.resetcred.ResetPassword` sets the *UPDATE_PASSWORD* RA on the session, this will necessarily come _after_ my required actions set on the *UserModel*.

After trying a quick patch to have `ResetPassword` set the RA on the *UserModel* instead of *AuthenticationSessionModel* and lean on the priority ordering of RAs, it seems that `ActionToken`s have a divergent execution flow from the one I was expecting using `org.keycloak.services.managers.AuthenticationManager#sortRequiredActionsByPriority` to sort the RAs by priority. The patch below remedies that.

I'm hoping some maintainers can comment on whether there are better ways to resolve this without requiring a fork?

```diff
Index: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java b/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java
--- a/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java	(revision cce8ee2b1ab7c925bfbb62d8f12a33e93d943040)
+++ b/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java	(date 1674084435068)
@@ -1082,7 +1082,9 @@
 
         evaluateRequiredActionTriggers(session, authSession, request, event, realm, user);
 
-        Optional<String> reqAction = user.getRequiredActionsStream().findFirst();
+        Stream<String> actions = sortRequiredActionsByPriority(realm, user.getRequiredActionsStream())
+                .map(RequiredActionProviderModel::getProviderId);
+        Optional<String> reqAction = actions.findFirst();
         if (reqAction.isPresent()) {
             return reqAction.get();
         }
```

**Reply by mrpatrickpilch** [reply]
Additional context:

https://groups.google.com/g/keycloak-dev/c/MqpYd7kKa3U

https://github.com/keycloak/keycloak/issues/16873

**Reply by OisCircle** [reply]
Those changes only resolved the order of user required actions, should we consider about the order of auth session required actions?
If so, the changes should applies to the codes below. (After your code)

```diff
if (!authSession.getRequiredActions().isEmpty()) {
-    return authSession.getRequiredActions().iterator().next();
+    return sortRequiredActionsByPriority(realm, authSession.getRequiredActions().stream())
+            .findFirst()
+            .map(RequiredActionProviderModel::getProviderId)
+            .orElse(authSession.getRequiredActions().iterator().next());
}
```

**Comment by mrpatrickpilch** [comment, marked as answer]
Settled on just changing the `Reset Password` action in the reset credentials flow. Instead of an authenticator that just adds the `UpdatePassword` required action (prompting the whole ordering difficulty), I just created an authenticator that mimics the `UpdatePassword` RA behavior and proceeds to finish the flow once the password is reset. Then the user is authenticated and proceeds to the required actions.

The simplest solutions are the best solutions.
